File: src/main/scala/us/zoom/data/dfence/providers/snowflake/models/PlaybookGrant.scala
package us.zoom.data.dfence.providers.snowflake.models

import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeObjectType
import us.zoom.data.dfence.providers.snowflake.models.opaque.GrantPrivilege

// Typesafe version of PlaybookPrivilegeGrant
case class PlaybookGrant(
  objectType: SnowflakeObjectType,
  pattern: PlaybookPattern,
  privileges: List[GrantPrivilege],
  grantType: PlaybookGrantType,
  enable: Boolean
)

File: src/main/java/us/zoom/data/dfence/providers/snowflake/revoke/models/PlaybookGrant.java
package us.zoom.data.dfence.providers.snowflake.revoke.models;

import com.google.common.collect.ImmutableList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeObjectType;
import us.zoom.data.dfence.providers.snowflake.revoke.models.wrappers.GrantPrivilege;
import us.zoom.data.dfence.providers.snowflake.validations.playbook.pattern.models.ResolvedPlaybookPattern;

public record PlaybookGrant(
    SnowflakeObjectType objectType,
    PlaybookPattern pattern,
    List<GrantPrivilege> privileges,
    ResolvedPlaybookPattern resolvedPattern,
    boolean enable) {
  public PlaybookGrant {
    privileges =
        ImmutableList.copyOf(
            privileges.stream().filter(Objects::nonNull).collect(Collectors.toList()));
  }
}
